---
title: "Analyze the diauxic shift expression data using K-means and hierarchical clustering"
author: "Bin He"
date: "2023-06-22 (updated `r Sys.Date()`)"
output: 
  html_notebook:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r load_library, echo=FALSE}
# install the package if not already available
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(ggtext))
suppressPackageStartupMessages(library(factoextra))
suppressPackageStartupMessages(library(clusterProfiler))
#suppressPackageStartupMessages(library(edgeR))
#suppressPackageStartupMessages(library(SummarizedExperiment))
#suppressPackageStartupMessages(require(broom))
```

## Goal
Learn to use K-means clustering on the diauxic shift expression time course.

## Data
The data are from DeRisi et al. 1997 (DOI: 10.1126/science.278.5338.680). Briefly, researchers sampled yeast grown in the YPD media at 2 hour intervals during a regular growth cycle, labeled the cDNA with Cy3. They used a common reference - the cDNA from the 0 time point, labeled with Cy5. By mixing the latter time point samples each with a copy of the 0 min sample, they are able to get the ratio of expression for each gene, which is presented in log2 transformed forms here.

_Data source_

| GEO# | Description | Reference |
| ---- | ----------- | --------- |
| GSE28 | Microarray expression patterns of _S. cerevisiae_ wt during the metabolic shift from anaerobic fermentation to aerobic respiration. Samples taken at several time points up to 20.5 h. | DeRisi et al. 1997 |

The analysis below is inspired by a GEOquery workshop offered by Jason Ratcliff at the Iowa Institute of Human Genetics. `getGEO()` will return an ExpressionSet object, which is then converted into a "SummarizedExperiment" object, which is a more modern data structure that is easier to deal with.
```
Sys.setenv("VROOM_CONNECTION_SIZE" = 131072*10) # increase the local cache size
gse <- GEOquery::getGEO(GEO = "GSE28", destdir = "../input/")
gse <- as(object = gse[[1]], "SummarizedExperiment")
```

_Alternative source_

Instead of processing the GSE data above, I downloaded an expression matrix from this [link](https://www.yeastgenome.org/reference/S000045635) and lightly processed it to remove unnecessary columns and rows, and renamed the column headers
```{r}
raw <- read_tsv("../input/20230622-derisi-1997-expression-matrix.tsv")
```

We will filter this dataset by excluding rows where all time point measurements are within -$\delta$ to $\delta$, with duplicate genes removed
```{r}
delta = 2.2
mat0 = as.matrix(select(raw, -ORF))
not.use = apply(mat0, 1, function(x) all(x > -delta & x < delta) || any(is.na(x)))
use = which(!not.use)
mat = mat0[use,]
rownames(mat) = raw$ORF[use]
all(!duplicated(rownames(mat)))
sprintf("We have a total of %d genes after filtering", nrow(mat))
```
There are no duplicated genes in the filtered set.

## Exploratory Analysis
Let's visualize the data
```{r}
dat.long <- raw[use,] %>% 
  pivot_longer(cols = -ORF, names_to = "time", values_to = "log2") %>% 
  mutate(time = gsub("t", "", time) %>% as.numeric())
```
```{r}
ggplot(dat.long, aes(x = time, y = log2)) + 
  geom_line(aes(group = ORF), linewidth = 0.5, alpha = 0.2)
```
## K-means clustering
We will use the `kmeans()` function in the `stats` package.
```{r}
# set random seed
set.seed(123)
res <- kmeans(mat, centers = 6, iter.max = 100)
```

Let's check out the clusters
```{r}
kmeans.labels = tibble(ORF = names(res$cluster), cluster = res$cluster)
left_join(dat.long, kmeans.labels, by = "ORF") %>%
  filter(!is.na(cluster)) %>% 
  ggplot(aes(x = time, y = log2)) + 
  geom_line(aes(group = ORF), linewidth = 0.5, alpha = 0.2) +
  facet_wrap(~cluster) + 
  theme_cowplot()
```

Exercise Break: Pick your favorite parameter k and run the Lloyd algorithm 1,0000 times on the 260-gene diauxic shift dataset, each time initialized with a new set of k randomly chosen centers. Construct a histogram of the squared error distortions of the resulting 1,000 outcomes. How many times did you have to run the Lloyd algorithm before finding the run that scored highest among your 1,000 runs?
```{r}
# for this test, we will specify the Lloyd algorithm. the default used by the 
#`kmeans` function is the "Hartigan and Wong 1979" algorithm
res <- sapply(1:1000, function(i){
  tmp <- kmeans(mat, centers = 6, iter.max = 100, nstart = 1, algorithm = "Lloyd")
  return(tmp$tot.withinss/nrow(mat))
})
h <- hist(res, breaks = 30, xlab = "Mean within cluster squared error",
          main = "Histogram of the squared error distortion from 1000 runs",
          sub = paste0("The run with the lowest error distortion: ", which.min(res)))
```

## Determining "K" using the Elbow method
The idea is to measure the total within cluster distortion over a range of K and visualize the relationship between the two. We expect decreasing total within cluster distortion with increasing K. The bent in the curve, i.e., "elbow", is usually where we would like to pick k
```{r}
# choose k between 2 to 10
wss_values <- sapply(1:10, function(i){
  kmeans(mat, centers = i, nstart = 10)$tot.withinss
})
```

```{r}
tibble(k = 1:10, wss_values) %>% 
  ggplot(aes(k, wss_values)) + geom_point(size = 3) + geom_line() +
  ylab("Total within-cluster sum of distortion") + 
   xlab("Number of clusters: k") +
  scale_x_continuous(breaks = 1:10) +
  theme_cowplot()
```

```{r}
factoextra::fviz_nbclust(mat, kmeans, method = "wss")
```

## Hierarchical clustering
Use `dist` to convert the expression data into a distance (dissimilarity) matrix. Two different [distance measures](https://uc-r.github.io/kmeans_clustering#distance) will be used, i.e., Euclidean distance (length of a line segment between two points) and Pearson's correlation distance (correlation). The quote below gives a good explanation for which one is used for what. ([reference](https://www.biostars.org/p/460492/))

> A good point to rise is data distribution importance for choosing distance measures in clustering analysis. Thanks This is my understanding of differences between Euclidean distance or Pearson correlation distances application for gene expression clustering: When we are interested in considering overall expression profiles (up and down), correlation-based measures (i.e. Pearson correlation) would be of choice. In other cases, we may want to cluster observations with the same magnitude of dysregulation together. In this way observations with high value of features would cluster together. In these cases, Euclidean distance would be our choice for dissimilarity matrix calculation.

```{r}
d.euclidean <- get_dist(mat, method = "euclidean")
d.pearson <- get_dist(mat, method = "pearson")
```

Perform hierarchical clustering
```{r}
#hc1 <- hclust(d.euclidean, method = "average")
hc1 <- hcut(d.euclidean, k = 6, hc_func = "hclust", hc_method = "average", graph = TRUE)
hc2 <- hcut(d.pearson, k = 6, hc_func = "hclust", hc_method = "average", graph = TRUE)
```

Plot the resulting dendrogram
```{r}
plot(hc1, cex = 0.1, hang = -1)
rect.hclust(hc1, k = 6, border = 2:7)
plot(hc2, cex = 0.1, hang = -1)
rect.hclust(hc2, k = 6, border = 2:7)
```

Use the Euclidean distance based clustering labels to show the clustering result (in terms of the expression patterns in each cluster)
```{r}
hc1.labels = tibble(ORF = names(hc1$cluster), cluster = hc1$cluster)
hc2.labels = tibble(ORF = names(hc2$cluster), cluster = hc2$cluster)
left_join(dat.long, hc1.labels, by = "ORF") %>%
  filter(!is.na(cluster)) %>% 
  ggplot(aes(x = time, y = log2)) + 
  geom_line(aes(group = ORF), linewidth = 0.5, alpha = 0.2) +
  facet_wrap(~cluster) + 
  theme_cowplot()
```

Comparing K-means and Hierarchical clustering
```{r}
all(hc.labels$ORF == kmeans.labels$ORF) # check to make sure the orders are the same
seq.grp <- cbind("HC_Euclidean" = hc1.labels$cluster,
                 "HC_Pearson" = hc2.labels$cluster,
                 "K-Means" = kmeans.labels$cluster)
plot(hc1, cex = 0.1, hang = -1)
dendextend::colored_bars(seq.grp, as.dendrogram(hc1), add = TRUE, y_shift = -0.1)
```

Let's compare the three clustering result using the [Rand Index](https://en.wikipedia.org/wiki/Rand_index)
```{r}
# from the package `fossil`, copied from 
# https://davetang.org/muse/2017/09/21/the-rand-index/
rand.index <- function (group1, group2) 
{
    x <- abs(sapply(group1, function(x) x - group1))
    x[x > 1] <- 1
    y <- abs(sapply(group2, function(x) x - group2))
    y[y > 1] <- 1
    sg <- sum(abs(x - y))/2
    bc <- choose(dim(x)[1], 2)
    ri <- 1 - sg/bc
    return(ri)
}
```

```{r}
RI <- c(
  "HC_Euclidean vs HC_Pearson" = rand.index(hc1.labels$cluster, hc2.labels$cluster),
  "HC_Euclidean vs K-Means" = rand.index(hc1.labels$cluster, kmeans.labels$cluster),
  "HC_Pearson vs K-Means" = rand.index(hc2.labels$cluster, kmeans.labels$cluster)
)
print(RI)
```

Export the cluster labels
```{r}
as.data.frame(seq.grp) %>% 
  rownames_to_column(var = "ORF") %>% 
  as_tibble() %>% 
  write_tsv(file = "../output/20230726-Kmeans-HC-cluster-labels.tsv")
```

Take some example gene sets to play with
```{r}
# try cluster 1 from Hierarchical clustering with Euclidean distance
test.set <- filter(hc1.labels, cluster == 1) %>% pull(ORF)
```

## Enrichment analysis
One of the purposes of clustering genes is to be able to identify shared biological processes and functions among genes with similar expression patterns. One way to do so is to use the controlled vocabulary such as Gene Ontology (GO). Many online sites exist to analyze yeast gene sets, such as [YeastMine](https://yeastmine.yeastgenome.org/yeastmine/), [EnrichR](https://maayanlab.cloud/YeastEnrichr/enrich), [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost). In this notebook, we will learn to use [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) (book manual available [here](https://yulab-smu.top/biomedical-knowledge-mining-book/clusterprofiler-go.html)).
```{r}
geneList <- read_csv("../output/20230726-example-gene-set-from-HC-entrezgene-id.csv")
```

```{r}
ggo <- enrichGO(gene = geneList$converted_alias,
                OrgDb = 'org.Sc.sgd.db',
                keyType = "ENTREZID",
                ont = "MF",
                readable = FALSE)
head(ggo)
```

